### ISSUE TITLE
[$170] Summarized TaaS Intake Form Changes

### ISSUE THREAD

**Issue by acshields** [appirio-tech/connect-app#4206]
@RishiRajSahu - the TaaS team and will have requested these updates be made to the TaaS intake form. Please review - if you have questions, clarify with Will directly :) 

https://docs.google.com/presentation/d/1jQFX-eAMjGn5D6uCp18OKmivW9FfYXd8vjTTWOxEUg4/edit#slide=id.p

**Comment by acshields** [comment]
@RishiRajSahu - do you think these changes can be accomplished before 1/15? If not, I will deprioritize from January for now (re: our new "project proposal" process that is kicking off in Prod/Dev at the start of the new year).

**Comment by maxceem** [comment]
@RishiRajSahu if I understand it correctly, from the code point of view we have to create a new component for entering Jobs instead of Talents as per the description of these 2 screens and the rest can be done by editing Project Template in Metadata:

     ![image](https://user-images.githubusercontent.com/146016/103518549-a371fd80-4e7c-11eb-9962-2e9245e3f35c.png)

     ![image](https://user-images.githubusercontent.com/146016/103518568-acfb6580-4e7c-11eb-8389-4284fc2cd1a6.png)

**Comment by maxceem** [comment]
@RishiRajSahu one question I have: should we remove the existing big field for **Job Description**, as we are adding another **Job Description** field on the first line?

**Comment by RishiRajSahu** [comment]
@maxceem yes let's remove it and also remove optional skills field too.

@acshields we are removing the redundant job-description field from the component.

**Comment by maxceem** [comment]
@RishiRajSahu do we have any ProjectTemplate which you would like to use on DEV for this?

I think we would name a new component for the jobs as `job-picker` (`JobPickerQuestion`) instead of `talent-picker`/`talent-picker-2` (`TalentPickerQuestion`/`TalentPickerQuestionV2`).

**Comment by RishiRajSahu** [comment]
@maxceem feel free to duplicate the one we are using in dev for taas.

**Comment by maxceem** [comment]
Sum up:
- At the moment we have a form for selecting talents https://connect.topcoder-dev.com/new-project/talent-as-a-service which looks like this when we are selecting talents:

   ![image](https://user-images.githubusercontent.com/146016/103640449-b5c16980-4f58-11eb-8b74-c39f8c5137f4.png)
   
   - This question for selecting talents has type [talent-picker-v2](https://github.com/appirio-tech/connect-app/blob/feature/taas-jobs-2/src/projects/detail/components/SpecQuestions.jsx#L386) is rendered by component [TalentPickerQuestionV2](https://github.com/appirio-tech/connect-app/blob/feature/taas-jobs-2/src/projects/detail/components/TalentPickerQuestion/TalentPickerQuestionV2.jsx#L181-L182)
   
- We already created a new form for selecting jobs https://connect.topcoder-dev.com/new-project/taas-jobs and we created a new configuration for the component for selecting jobs. We need to create a new component `JobsPickerQuestion` to render type `jobs-picker`:

    ![image](https://user-images.githubusercontent.com/146016/103640570-e73a3500-4f58-11eb-9c0f-aa7a862be141.png)

- From the technical point of view we can copy component `TalentPickerQuestionV2` and make changes to it as per the description below.

- Reuse existent UI elements.
   
![image](https://user-images.githubusercontent.com/146016/103641175-fc639380-4f59-11eb-9664-f9757682cb8a.png)

![image](https://user-images.githubusercontent.com/146016/103641204-0daca000-4f5a-11eb-9e71-86a69ce4f260.png)

**Comment by maxceem** [comment]
Contest https://www.topcoder.com/challenges/30161295 has been created for this ticket.<br/><br/>```This is an automated message for maxceem via Topcoder X```

**Comment by maxceem** [comment]
@yoution @nqviet this is open for pick up.

**Comment by yoution** [comment]
@maxceem  please assign to me

**Comment by maxceem** [comment]
Contest https://www.topcoder.com/challenges/30161295 has been updated - it has been assigned to yoution.<br/><br/>```This is an automated message for maxceem via Topcoder X```

**Comment by maxceem** [comment]
@yoution assigned. 

One important thing, we have to use another branch for this issue: `feature/taas-jobs-2`.

If you have any questions, please, let me know.

**Comment by RishiRajSahu** [comment]
@acshields regarding this https://github.com/appirio-tech/connect-app/issues/4206#issuecomment-753911001, Could you please confirm that we need description field for jobs ? Its not clear in the slides whether -
1. We need to have a name field with default value of project name and could be updated with description of job by user ? or
2. Name field and description fields are separate fields, and name field will have default value of project-name

In case of 2, we would have to have a minor change on Taas-api backend which should easily doable.

**Comment by RishiRajSahu** [comment]
@acshields  thanks, as a summary for @wdprice, we need a field `job_name` to be defined for jobs object. fyi @maxceem

**Comment by maxceem** [comment]
@yoution just for you to start more easily. Basically the next steps should be done:
- Copy `TalentPickerQuestionV2` to `JobsPickerQuestion`
- Use it to render question type `jobs-picker` here https://github.com/appirio-tech/connect-app/blob/feature/taas-jobs-2/src/projects/detail/components/SpecQuestions.jsx#L386
- After that in this component make the next changes:
   - show only one component instead many components for each job Designer, Developer, QA and so on
   - remove icon with text like Designer/Developer
   - Add "Job Title" field on the first line (empty by default) it should be required
   - Add "Job Role" select dropdown
   - Remove "Prefered Good to have skills" skills select (so we keep only one required field for skills)
   - Verification should work the same as now:
      - at least one job should have all required fields filled
      - if we add a second job, then it should also have all required fields filled, and so on
      
Full requirements are in this comment https://github.com/appirio-tech/connect-app/issues/4206#issuecomment-754580004

**Comment by maxceem** [comment]
@yoution as per my testing all good with this task. The only thing we should not send `_key` to the server, we add this property only for internal usage to avoid rendering issues:

![image](https://user-images.githubusercontent.com/146016/103887818-601dc600-50ec-11eb-86a9-e4a086dbbe1a.png)

Could you please, create a new PR so we don't send this to server. Please, pull the latest changes from the branch first, as I did some updates there.

**Comment by yoution** [comment]
@maxceem please review

**Comment by maxceem** [comment]
@yoution I found one issue in the form, which is more important than sending `_key`.

When I start typing the first job title, after the first letter is typed, the field loose the focus and I cannot type anymore. I have click the job title field again, to continue typeing, see demo video https://monosnap.com/file/9GG8eHLPCkuionVQOUw6NebXD3dZ5W.

You may also test it online here https://connect-test01.topcoder-dev.com/new-project/taas-jobs

**Comment by yoution** [comment]
@maxceem have pushed, please reivew, I also find a new method to store the `_key`, like so ,store in the prototype, 
![image](https://user-images.githubusercontent.com/3863914/103970488-530fde00-51a3-11eb-9be3-16aae04a3eed.png), 

but currently I have move the values to state, so donot need so complex

**Comment by maxceem** [comment]
Thank you @yoution. I would review it in some time. 

Did you have a chance to find the reason for Job Title loosing focus during typing https://github.com/appirio-tech/connect-app/issues/4206#issuecomment-756162413 ?

**Comment by yoution** [comment]
@maxceem when you change the title, the `_key` will regenerate a new one, the the key wil be not the same, so the component will unmount and create a new one
https://github.com/appirio-tech/connect-app/blob/feature/taas-jobs-2/src/projects/detail/components/JobsPickerQuestion/JobsPickerQuestion.jsx#L74

please try this demo
https://codesandbox.io/s/react-otp-input-demo-forked-nwuew

but I don't know why this happen, when I test locally, this so obviously problem I will pay attention to

**Comment by maxceem** [comment]
Hmm, yeah, `_key` should be generated only once per record.

**Comment by yoution** [comment]
@maxceem this demo,  https://codesandbox.io/s/mystifying-browser-5hgrs?file=/src/App.js:0-478

**Comment by maxceem** [comment]
Yeah, if `_key` got updated the focus would be lost.
So we have to make sure, that `_key` is always the same, not get updated.

**Comment by yoution** [comment]
have pushed

**Comment by maxceem** [comment]
@drasticdpk this is ready for partial testing.

- We've implemented new component for entering jobs as per the description in this comment https://github.com/appirio-tech/connect-app/issues/4206#issuecomment-754580004
- So the part inside the next component is ready for testing:

   ![image](https://user-images.githubusercontent.com/146016/104091619-2e8d3200-5287-11eb-8900-05604567a8f6.png)

**Comment by drasticdpk** [comment]
- User should not able to put space char in text box which is mandatory fields. 
![image](https://user-images.githubusercontent.com/25297859/104292863-727d7280-54e5-11eb-96a1-2faf92327a92.png)

- If there is empty job container, Next button should be disable. 
![image](https://user-images.githubusercontent.com/25297859/104293270-f0da1480-54e5-11eb-9977-95f8b61dfdb2.png)

@maxceem  please suggest.

**Comment by maxceem** [comment]
good suggestions @drasticdpk

@yoution could you please, handle these issues? https://github.com/appirio-tech/connect-app/issues/4206#issuecomment-758515990

I would increase the prize for this issue for $20.

Regarding:
> User should not able to put space char in text box which is mandatory fields.

I can see that we already ignore the spaces in other fields like the project title so we should be able to config it for this field also:

![image](https://user-images.githubusercontent.com/146016/104298266-6e912680-54cc-11eb-8ba5-72fa3608e39b.png)

regarding:
> If there is empty job container, Next button should be disable.

I know this work like this in the previous form, but the logic there was lightly different, so in this new form, we should not allow adding empty job.

**Comment by yoution** [comment]
@maxceem please review

**Comment by acshields** [comment]
@maxceem @RishiRajSahu - Mess saw the updated form in Dev and made one small change request: Spilt Hiring Manager Name and Hiring Manager Email into separate fields.

![Screen Shot 2021-01-12 at 9 48 15 AM](https://user-images.githubusercontent.com/41118149/104331143-5ca38980-54b4-11eb-94c1-f4a2c157bc1e.png)

Updates:
Field 1 Label: Hiring Manager Name
Field 1 Answer Format: Text Input Field

Field 2 Label: Hiring Manager Email
Field 2 Answer Format: Text Input Field

Rules around who has visibility to these fields have not changed.

**Comment by maxceem** [comment]
@yoution the PR fixes the first issue "User should not able to put space char in text box which is mandatory fields." so I've merged it.

But there is a second part:
> If there is empty job container, Next button should be disable.

![image](https://user-images.githubusercontent.com/146016/104338157-df513680-54fe-11eb-986b-cc822a654240.png)

**Comment by yoution** [comment]
@maxceem please reivew

**Comment by maxceem** [comment]
@drasticdpk both these issues https://github.com/appirio-tech/connect-app/issues/4206#issuecomment-758515990 were fixed and ready for QA via https://connect-test01.topcoder-dev.com/new-project/taas-jobs.

**Comment by drasticdpk** [comment]
Verified. 
![image](https://user-images.githubusercontent.com/25297859/104460394-28c08500-55d6-11eb-8cea-5aa5ef55ee33.png)
![image](https://user-images.githubusercontent.com/25297859/104460556-57d6f680-55d6-11eb-95fd-2d0f3e227bce.png)

**Comment by maxceem** [comment]
Payment processed via https://www.topcoder.com/challenges/826a53b6-3981-496e-8c91-0b27a386fd28

**Comment by drasticdpk** [comment]
Verified on Prod
![image](https://user-images.githubusercontent.com/25297859/106143267-a4a3fb00-619a-11eb-8f40-4a401a04d9ca.png)

@maxceem however below change request has not implemented yet. Shall I create it new issue ? 
> https://github.com/appirio-tech/connect-app/issues/4206#issuecomment-758711905 

Reference : 
![image](https://user-images.githubusercontent.com/25297859/106143664-3449a980-619b-11eb-85cd-6c6ebda0d630.png)

**Comment by maxceem** [comment]
This ticket was not processed for payment. If you would like to process it for payment, please reopen it, add the ```tcx_FixAccepted``` label, and then close it again<br/><br/>```This is an automated message for maxceem via Topcoder X```
